### ISSUE TITLE
gocritic panics when there is a method with Go generic type receiver 

### ISSUE THREAD

**Issue by senpathi** [golangci/golangci-lint#3357]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc.).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

In my main/main.go file, there is a `go generic struct` and when it has a `method with generic receiver`.
I tried to run `golangci-lint run` with the `gocritic linter` for this file and it panics with following error.
```
ERRO Running error: 1 error occurred:
        * can't run linter gocritic: gocritic: gocritic: package "main" (isInitialPkg: true, needAnalyzeSource: true): go/types/sizes.go:82: assertion failed
``` 
If Generic type receiver removed, there is no error.

### Version of golangci-lint

<details>

```console
golangci-lint has version 1.50.1 built from 8926a95 on 2022-10-22T10:48:48Z
```

</details>

### Configuration file

<details>

```yaml
linters-settings:
  gocritic:
    # Which checks should be enabled; can't be combined with 'disabled-checks'.
    enabled-tags:
      - diagnostic
      - experimental
      - opinionated
      - performance
      - style

linters:
  disable-all: true
  enable:
    - gocritic

# Options for analysis running.
run:
  # Timeout for analysis, e.g. 30s, 5m.
  timeout: 5m
```
</details>

### Go environment

<details>

```console
go version go1.18.3 darwin/amd64
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/nipunapathirana/Library/Caches/go-build"
GOENV="/Users/nipunapathirana/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/nipunapathirana/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/nipunapathirana/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.18.3"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/nipunapathirana/go/src/testProject1/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/bj/n3h9c3790ms7qtypvw3lsknw0000gn/T/go-build2686105363=/tmp/go-build -gno-record-gcc-switches -fno-common"


```

</details>

### Verbose output of running

<details>

```console
INFO [config_reader] Config search paths: [./ /Users/nipunapathirana/go/src/testProject1 /Users/nipunapathirana/go/src /Users/nipunapathirana/go /Users/nipunapathirana /Users /] 
INFO [config_reader] Used config file .golangci.yaml 
INFO [lintersdb] Active 1 linters: [gocritic]     
INFO [loader] Go packages loading at mode 575 (name|types_sizes|compiled_files|deps|exports_file|files|imports) took 451.779959ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 1.962833ms 
INFO [linters_context/goanalysis] analyzers took 1.999472334s with top 10 stages: gocritic: 1.999472334s 
ERRO [runner] Panic: gocritic: package "main" (isInitialPkg: true, needAnalyzeSource: true): go/types/sizes.go:82: assertion failed: goroutine 289 [running]:
runtime/debug.Stack()
        runtime/debug/stack.go:24 +0x64
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func1()
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:102 +0x104
panic({0x1031bc3c0, 0x1400360d2f0})
        runtime/panic.go:884 +0x204
go/types.assert(0x88?)
        go/types/errors.go:27 +0x60
go/types.(*StdSizes).Alignof(0x140000491d0, {0x10338d288?, 0x14000f47380})
        go/types/sizes.go:82 +0x1a8
go/types.(*StdSizes).Offsetsof(0x10338d0f8?, {0x140002172f0, 0x1, 0x0?})
        go/types/sizes.go:123 +0xc8
go/types.(*StdSizes).Sizeof(0x140000491d0, {0x10338d0f8?, 0x1400037bdc0})
        go/types/sizes.go:176 +0x278
github.com/go-critic/go-critic/framework/linter.(*CheckerContext).SizeOf(...)
        github.com/go-critic/go-critic@v0.6.5/framework/linter/linter.go:325
github.com/go-critic/go-critic/checkers.(*hugeParamChecker).checkParams(0x140013d9470, {0x14000217248, 0x1, 0x103348ee0?})
        github.com/go-critic/go-critic@v0.6.5/checkers/hugeParam_checker.go:56 +0x11c
github.com/go-critic/go-critic/checkers.(*hugeParamChecker).VisitFuncDecl(0x1400154ad08?, 0x14000f46de0)
        github.com/go-critic/go-critic@v0.6.5/checkers/hugeParam_checker.go:44 +0x38
github.com/go-critic/go-critic/checkers/internal/astwalk.(*funcDeclWalker).WalkFile(0x14001544e50, 0x1400013f500)
        github.com/go-critic/go-critic@v0.6.5/checkers/internal/astwalk/func_decl_walker.go:21 +0xec
github.com/go-critic/go-critic/framework/linter.(*Checker).Check(...)
        github.com/go-critic/go-critic@v0.6.5/framework/linter/linter.go:130
github.com/golangci/golangci-lint/pkg/golinters.runGocriticOnFile(0x14000ef1e60, 0x14000a6cf2b?, {0x14007d2f000, 0x69, 0x10322e5a0?})
        github.com/golangci/golangci-lint/pkg/golinters/gocritic.go:175 +0xa0
github.com/golangci/golangci-lint/pkg/golinters.runGocriticOnPackage(0x14000ef1e60, {0x14007d2f000, 0x69, 0x80}, {0x14000217208, 0x1, 0x14002261cb8?})
        github.com/golangci/golangci-lint/pkg/golinters/gocritic.go:163 +0x15c
github.com/golangci/golangci-lint/pkg/golinters.(*goCriticWrapper).run(0x14000f6ac00, 0x14000ed1e10)
        github.com/golangci/golangci-lint/pkg/golinters/gocritic.go:123 +0x1c8
github.com/golangci/golangci-lint/pkg/golinters.NewGoCritic.func1(0x103226140?)
        github.com/golangci/golangci-lint/pkg/golinters/gocritic.go:46 +0x38
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyze(0x14000130d80)
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:188 +0x958
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func2()
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:106 +0x20
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0x140012aea50, {0x102e9bbca, 0x8}, 0x140003e5730)
        github.com/golangci/golangci-lint/pkg/timeutils/stopwatch.go:111 +0x44
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe(0x23?)
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:105 +0x74
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze.func2(0x14000130d80)
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_loadingpackage.go:80 +0xb0
created by github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_loadingpackage.go:75 +0x168 
WARN [runner] Can't run linter gocritic: gocritic: gocritic: package "main" (isInitialPkg: true, needAnalyzeSource: true): go/types/sizes.go:82: assertion failed 
INFO [runner] processing took 2.126µs with stages: max_same_issues: 542ns, skip_dirs: 292ns, max_from_linter: 250ns, nolint: 208ns, cgo: 167ns, filename_unadjuster: 167ns, skip_files: 167ns, identifier_marker: 42ns, source_code: 42ns, path_prefixer: 42ns, sort_results: 42ns, path_prettifier: 42ns, exclude: 41ns, uniq_by_line: 41ns, exclude-rules: 41ns, path_shortener: 0s, severity-rules: 0s, max_per_file_from_linter: 0s, diff: 0s, autogenerated_exclude: 0s 
INFO [runner] linters took 2.128246375s with stages: gocritic: 2.128059084s 
ERRO Running error: 1 error occurred:
        * can't run linter gocritic: gocritic: gocritic: package "main" (isInitialPkg: true, needAnalyzeSource: true): go/types/sizes.go:82: assertion failed
 
INFO Memory: 27 samples, avg is 102.1MB, max is 192.8MB 
INFO Execution took 2.599534125s 
```

</details>

### Code example or link to a public repository

<details>

```go
package main

import "fmt"

type Test[T any] struct {
	ABC T `json:"abc"`
}

func (req Test[T]) Print() {
	fmt.Println("ABC")
}

func main() {
}
```

</details>

**Comment by ldez** [comment]
Hello,

have you tried the [go-critic](https://github.com/go-critic/go-critic) command line as suggested by the checkbox "Yes, I've tried with the standalone linter if available"?

**Comment by senpathi** [comment]
Yes, 
It also panicked, Output below.
### Command
```
gocritic check -enableAll ./...
```


### Output
<Details>
```
panic: assertion failed [recovered]
        panic: assertion failed

goroutine 1582 [running]:
github.com/go-critic/go-critic/framework/lintmain/internal/check.(*program).checkFile.func1.1()
        /Users/nipunapathirana/go/pkg/mod/github.com/go-critic/go-critic@v0.6.5/framework/lintmain/internal/check/check.go:160 +0x11e
panic({0x43fefe0, 0x4522d10})
        /usr/local/go/src/runtime/panic.go:838 +0x207
go/types.assert(...)
        /usr/local/go/src/go/types/errors.go:21
go/types.(*StdSizes).Alignof(0xc00023c210, {0x4525b30?, 0xc001068930})
        /usr/local/go/src/go/types/sizes.go:72 +0x25f
go/types.(*StdSizes).Offsetsof(0x4525a18?, {0xc0006160d0, 0x1, 0x0?})
        /usr/local/go/src/go/types/sizes.go:101 +0xdd
go/types.(*StdSizes).Sizeof(0xc00023c210, {0x4525a18?, 0xc00024a070})
        /usr/local/go/src/go/types/sizes.go:154 +0x1c5
github.com/go-critic/go-critic/framework/linter.(*CheckerContext).SizeOf(...)
        /Users/nipunapathirana/go/pkg/mod/github.com/go-critic/go-critic@v0.6.5/framework/linter/linter.go:325
github.com/go-critic/go-critic/checkers.(*hugeParamChecker).checkParams(0xc00103c348, {0xc000616048, 0x1, 0x0?})
        /Users/nipunapathirana/go/pkg/mod/github.com/go-critic/go-critic@v0.6.5/checkers/hugeParam_checker.go:56 +0x107
github.com/go-critic/go-critic/checkers.(*hugeParamChecker).VisitFuncDecl(0x0?, 0xc001068360)
        /Users/nipunapathirana/go/pkg/mod/github.com/go-critic/go-critic@v0.6.5/checkers/hugeParam_checker.go:44 +0x38
github.com/go-critic/go-critic/checkers/internal/astwalk.(*funcDeclWalker).WalkFile(0xc00026bcc0, 0xc000b84000)
        /Users/nipunapathirana/go/pkg/mod/github.com/go-critic/go-critic@v0.6.5/checkers/internal/astwalk/func_decl_walker.go:21 +0xda
github.com/go-critic/go-critic/framework/linter.(*Checker).Check(...)
        /Users/nipunapathirana/go/pkg/mod/github.com/go-critic/go-critic@v0.6.5/framework/linter/linter.go:130
github.com/go-critic/go-critic/framework/lintmain/internal/check.(*program).checkFile.func1(0x2b, 0xc0002f4500)
        /Users/nipunapathirana/go/pkg/mod/github.com/go-critic/go-critic@v0.6.5/framework/lintmain/internal/check/check.go:164 +0xc2
created by github.com/go-critic/go-critic/framework/lintmain/internal/check.(*program).checkFile
        /Users/nipunapathirana/go/pkg/mod/github.com/go-critic/go-critic@v0.6.5/framework/lintmain/internal/check/check.go:146 +0xa5
```

</Details>

**Comment by ldez** [comment]
So it's an issue with [gocritic](https://github.com/go-critic/go-critic/issues).
